File: backend/app/schemas/resume.py
"""
backend/app/schemas/resume.py

Pydantic schemas for the Resume entity.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ResumeAnalysisCreate(BaseModel):
    """Schema for the data needed to persist a completed resume analysis."""

    user_id: int
    original_filename: str
    stored_filename: str
    file_path: str
    file_size: int
    extracted_text: str
    ats_score: int
    summary: str
    technical_skills: list[str]
    soft_skills: list[str]
    missing_skills: list[str]
    strengths: list[str]
    weaknesses: list[str]
    improvement_suggestions: list[str]
    recommended_job_roles: list[str]
    overall_recommendation: str


class ResumeResponse(BaseModel):
    """Schema for returning full resume + analysis information in API responses."""

    id: int
    user_id: int
    original_filename: str
    stored_filename: str
    file_path: str
    file_size: int
    extracted_text: str
    ats_score: int
    summary: str
    technical_skills: list[str]
    soft_skills: list[str]
    missing_skills: list[str]
    strengths: list[str]
    weaknesses: list[str]
    improvement_suggestions: list[str]
    recommended_job_roles: list[str]
    overall_recommendation: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeSummaryResponse(BaseModel):
    """Lightweight schema for listing resumes without the full analysis payload."""

    id: int
    original_filename: str
    ats_score: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeUploadResponse(BaseModel):
    """Schema returned immediately after resume upload (before AI analysis)."""

    id: int
    user_id: int
    original_filename: str
    stored_filename: str
    file_path: str
    file_size: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/resume.ts
// src/types/resume.ts

/**
 * Matches the FastAPI Resume response model exactly.
 * Field names are kept in snake_case to mirror the backend JSON payload
 * as-is (no camelCase transformation layer), avoiding silent mismatches
 * with the API contract.
 */
export interface Resume {
  id: number;
  user_id: number;
  original_filename: string;
  stored_filename: string;
  file_path: string;
  file_size: number;
  created_at: string;
}

/**
 * Structured output of the AI resume analysis pipeline.
 */
export interface ResumeAnalysisResult {
  ats_score: number;
  summary: string;
  technical_skills: string[];
  soft_skills: string[];
  missing_skills: string[];
  strengths: string[];
  weaknesses: string[];
  improvement_suggestions: string[];
  recommended_job_roles: string[];
  overall_recommendation: string;
}

/**
 * Response returned immediately after a resume upload completes.
 * Extends the base Resume metadata; analysis may not be ready yet,
 * so it's optional until the AI pipeline finishes processing.
 */
export interface ResumeUploadResponse extends Resume {
  analysis?: ResumeAnalysisResult;
}

/**
 * Response returned by the POST /resume/analyze endpoint.
 *
 * Example:
 * {
 *   "message": "Resume analyzed successfully.",
 *   "analysis": {
 *     "ats_score": 85,
 *     "summary": "...",
 *     ...
 *   }
 * }
 */
export interface AnalyzeResumeResponse {
  message: string;
  analysis: ResumeAnalysisResult;
}